File: services/claims_scoring/src/attestation.rs
use anyhow::Result;
use tracing::{info, warn};

/// Verifies a Play Integrity API device attestation token.
///
/// In production, calls the Play Integrity API with the provided token.
/// In dev/test (when PLAY_INTEGRITY_API_KEY is not set), treats all tokens as valid
/// and logs a warning.
///
/// Returns `Ok(true)` if the device is attested, `Ok(false)` if attestation fails.
/// Returns `Err` if the attestation service itself is unreachable.
pub async fn verify_attestation(token: &str) -> Result<bool> {
    let api_key = std::env::var("PLAY_INTEGRITY_API_KEY").ok();

    match api_key {
        None => {
            warn!(
                "PLAY_INTEGRITY_API_KEY not set — treating all attestation tokens as valid (dev/test mode)"
            );
            Ok(true)
        }
        Some(key) => call_play_integrity_api(token, &key).await,
    }
}

async fn call_play_integrity_api(token: &str, api_key: &str) -> Result<bool> {
    // Play Integrity API endpoint
    let url = "https://playintegrity.googleapis.com/v1:decodeIntegrityToken";

    let client = reqwest::Client::new();
    let body = serde_json::json!({
        "integrity_token": token
    });

    let response = client
        .post(url)
        .query(&[("key", api_key)])
        .json(&body)
        .send()
        .await?;

    if !response.status().is_success() {
        tracing::error!(
            status = %response.status(),
            "Play Integrity API returned non-success status"
        );
        return Ok(false);
    }

    let payload: serde_json::Value = response.json().await?;

    // Check MEETS_DEVICE_INTEGRITY verdict
    let verdict = payload
        .pointer("/tokenPayloadExternal/deviceIntegrity/deviceRecognitionVerdict")
        .and_then(|v| v.as_array());

    match verdict {
        Some(verdicts) => {
            let attested = verdicts
                .iter()
                .any(|v| v.as_str() == Some("MEETS_DEVICE_INTEGRITY"));
            info!(attested, "Play Integrity API attestation result");
            Ok(attested)
        }
        None => {
            warn!("Play Integrity API response missing deviceRecognitionVerdict");
            Ok(false)
        }
    }
}

File: services/claims_scoring/attestation.py
import logging
import os

import httpx

logger = logging.getLogger(__name__)

_PLAY_INTEGRITY_URL = "https://playintegrity.googleapis.com/v1:decodeIntegrityToken"


async def verify_attestation(token: str) -> bool:
    api_key = os.getenv("PLAY_INTEGRITY_API_KEY")
    if not api_key:
        logger.warning("PLAY_INTEGRITY_API_KEY not set — accepting all tokens (dev/test mode)")
        return True
    return await _call_play_integrity_api(token, api_key)


async def _call_play_integrity_api(token: str, api_key: str) -> bool:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            _PLAY_INTEGRITY_URL,
            params={"key": api_key},
            json={"integrity_token": token},
        )

    if not response.is_success:
        logger.error("Play Integrity API returned status %d", response.status_code)
        return False

    payload = response.json()
    verdicts = (
        payload
        .get("tokenPayloadExternal", {})
        .get("deviceIntegrity", {})
        .get("deviceRecognitionVerdict", [])
    )
    attested = "MEETS_DEVICE_INTEGRITY" in verdicts
    logger.info("Play Integrity attestation result: %s", attested)
    return attested
